Generate slice.Tensor view operations instead of as_strided when split is used in the original program.

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137225

test_recompile assert that the changes do not add more recompilation by comparing with eager backend.
The reason of this is because slice can be lowered in more efficient way. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

laithsakka committed:
handle splot

[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -39,18 +50,57 @@ def regenerate_view(self, bases_list: List[Tensor]):
 
 
+# return None or (dim, start, end)
+def try_use_slice(base, tensor):
```
I'm concerned about what happens with dynamic shapes. As I mentioned in the other PR better testing would help convince me that it works out

laithsakka replied:
check test_recompiles added it in the previous should address this comment.

laithsakka committed:
Update on "Generate slice.Tensor view operations instead of as_strided when split is used in the original program. "




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Generate slice.Tensor view operations instead of as_strided when split is used in the original program. "




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Generate slice.Tensor view operations instead of as_strided when split is used in the original program. "




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Generate slice.Tensor view operations instead of as_strided when split is used in the original program. "




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Generate slice.Tensor view operations instead of as_strided when split is used in the original program. "


test_recompile assert that the changes do not add more recompilation by comparing with eager backend.
The reason of this is because slice can be lowered in more efficient way. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Generate slice.Tensor view operations instead of as_strided when split is used in the original program. "


test_recompile assert that the changes do not add more recompilation by comparing with eager backend.
The reason of this is because slice can be lowered in more efficient way. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Generate slice.Tensor view operations instead of as_strided when split is used in the original program. "


test_recompile assert that the changes do not add more recompilation by comparing with eager backend.
The reason of this is because slice can be lowered in more efficient way. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Generate slice.Tensor view operations instead of as_strided when split is used in the original program. "


test_recompile assert that the changes do not add more recompilation by comparing with eager backend.
The reason of this is because slice can be lowered in more efficient way. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison reviewed (commented):
looks good but deferring review to @zou3519 and @bdhirsh

zou3519 commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -28,9 +28,20 @@ def get_base(tensor):
-    size: Optional[Sequence[Union[int, torch.SymInt]]] = None
-    stride: Optional[Sequence[Union[int, torch.SymInt]]] = None
-    storage_offset: Optional[int] = None
+    # size, stride, and storage_offset passed to as_strided.
```
Can we make ViewInfo a Union[AsStridedInfo, SliceInfo, AliasInfo]? This class is getting a bit big

zou3519 replied:
They can all be subclasses of a base ViewInfo class that has a .regenerate_view function on it

zou3519 replied:
Maybe add a comment here for "If you need more complicated views in the future, then you're out of luck -- go implement auto_functionalized_v3"

laithsakka replied:
i was thinking for one more for subscript since its common

zou3519 replied:
Sure, that's fine by me

zou3519 commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -95,11 +155,11 @@ def write_single_view(prefix: str, tensor: Tensor, base_index: int):
-                kwargs[f"{prefix}_alias"] = True
+                use_alias()
+            elif (slice_info := try_use_slice(base, tensor)) is not None:
+                use_slice(*slice_info)
```
nit: could be methods on the ViewInfos if we do the refactor I mentioned above, but not a big deal

zou3519 commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -149,16 +209,22 @@ def read_single_view(prefix):
+                base_index, as_strided_info=(size, stride, storage_offset), is_view=True
+            )
+        elif f"{prefix}_slice_dim" in kwargs:
+            dim = get_arg(f"{prefix}_slice_dim")
```
(no action required) I also don't like how the serialization scheme is going. Maybe leave a comment that "if you want to support auto_functionalized_v2 for export -- please just implement auto_functionalized_v3"

zou3519 commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -47,18 +58,49 @@ def regenerate_view(self, bases_list: List[Tensor]):
+    ):
+        return (0, 0, base.size()[0])
+
+    # TODO can we relax this for better coverage?
```
what do you mean?

laithsakka replied:
does slice always preserve stride() and size()? I though not. 
I mean there might be other cases where we can regenerate slice

zou3519 replied:
slice() semantics are that it always returns a tensors with the same strides and it returns a tensor with the same sizes in all but one dimension. If we're talking about reconstructing the result of 2+ slices then... we're looking at auto_functionalized_v3

zou3519 reviewed (approved):
LGTM, but had some comments about refactoring the code to make it a bit nicer, please read

bdhirsh commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -95,11 +155,11 @@ def write_single_view(prefix: str, tensor: Tensor, base_index: int):
-                kwargs[f"{prefix}_size"] = tensor.size()
-                kwargs[f"{prefix}_stride"] = tensor.stride()
-                kwargs[f"{prefix}_storage_offset"] = tensor.storage_offset()
+                use_as_strided(tensor)
```
is there info somewhere on why `as_strided` is worse for us than `slice()` (for inductor?)

laithsakka replied:
so there is no benchmark to show, but talking to @eellison in lowering code, you can we realize(), which we do not on the slice path.

laithsakka commented:
address all comments

laithsakka committed:
Update on "Generate slice.Tensor view operations instead of as_strided when split is used in the original program. "


test_recompile assert that the changes do not add more recompilation by comparing with eager backend.
The reason of this is because slice can be lowered in more efficient way. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka commented:
@pytorchbot  merge

laithsakka committed:
Update on "Generate slice.Tensor view operations instead of as_strided when split is used in the original program. "


test_recompile assert that the changes do not add more recompilation by comparing with eager backend.
The reason of this is because slice can be lowered in more efficient way. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

laithsakka commented:
@pytorchbot merge
